web-common/src/features/project/ProjectDeployer.ts

```diff
@@ -0,0 +1,177 @@
+import { page } from "$app/stores";
+import type { ConnectError } from "@connectrpc/connect";
+import { extractDeployError } from "@rilldata/web-common/features/project/deploy-errors";
+import { queryClient } from "@rilldata/web-common/lib/svelte-query/globalQueryClient";
+import { waitUntil } from "@rilldata/web-common/lib/waitUtils";
+import { behaviourEvent } from "@rilldata/web-common/metrics/initMetrics";
+import { BehaviourEventAction } from "@rilldata/web-common/metrics/service/BehaviourEventTypes";
+import type { DeployValidationResponse } from "@rilldata/web-common/proto/gen/rill/local/v1/api_pb";
+import {
+  createLocalServiceDeploy,
+  createLocalServiceDeployValidation,
+  createLocalServiceRedeploy,
+  getLocalServiceGetCurrentUserQueryKey,
+  localServiceGetCurrentUser,
+} from "@rilldata/web-common/runtime-client/local-service";
+import { derived, get, writable } from "svelte/store";
+
+export class ProjectDeployer {
+  public readonly validation = createLocalServiceDeployValidation({
+    query: {
+      refetchOnWindowFocus: true,
+    },
+  });
+  public readonly promptOrgSelection = writable(false);
+
+  private readonly deploying = writable(false);
+  private readonly deployMutation = createLocalServiceDeploy();
+  private readonly redeployMutation = createLocalServiceRedeploy();
+
+  public get isDeployed() {
+    const validation = get(this.validation).data as DeployValidationResponse;
+    return !!validation?.deployedProjectId;
+  }
+
+  public getStatus() {
+    return derived(
+      [
+        this.validation,
+        this.deployMutation,
+        this.redeployMutation,
+        this.deploying,
+      ],
+      ([validation, deployMutation, redeployMutation, deploying]) => {
+        if (
+          validation.isFetching ||
+          deployMutation.isLoading ||
+          redeployMutation.isLoading ||
+          deploying
+        ) {
+          return {
+            isLoading: true,
+            error: undefined,
+          };
+        }
+
+        return {
+          isLoading: false,
+          error: extractDeployError(
+            (validation.error as ConnectError) ??
+              (deployMutation.error as ConnectError) ??
+              (redeployMutation.error as ConnectError),
+          ).message,
+        };
+      },
+    );
+  }
+
+  public async validate() {
+    let validation = get(this.validation).data as DeployValidationResponse;
+    if (validation?.deployedProjectId) {
+      return true;
+    }
+
+    await waitUntil(() => !get(this.validation).isFetching);
+    validation = get(this.validation).data as DeployValidationResponse;
+
+    if (!validation.isAuthenticated) {
+      void behaviourEvent?.fireDeployEvent(BehaviourEventAction.LoginStart);
+      window.open(
+        `${validation.loginUrl}/?redirect=${get(page).url.toString()}`,
+        "_self",
+      );
+      return false;
```

Comment:
It initially surprised me to see the `validate()` function opens up a log-in page as a side effect. Might be more clear if you moved this to the `deploy()` function.

Reply:
Ah the validate method is meant to read through DeployValidation and make sure it passes.